RoaringBitmap/src/main/java/org/roaringbitmap/BitSetUtil.java

```diff
@@ -14,14 +14,75 @@
  *
  */
 public class BitSetUtil {
-  // todo: add a method to convert a RoaringBitmap to a BitSet using BitSet.valueOf
 
   // a block consists has a maximum of 1024 words, each representing 64 bits,
   // thus representing at maximum 65536 bits
   public static final int BLOCK_LENGTH = BitmapContainer.MAX_CAPACITY / Long.SIZE; //
   // 64-bit
   // word
 
+  /**
+   * Convert a {@link RoaringBitmap} to a {@link BitSet}.
+   * <p>
+   * Equivalent to calling {@code BitSet.valueOf(BitSetUtil.toLongArray(bitmap))}.
+   */
+  public static BitSet bitsetOf(RoaringBitmap bitmap) {
+    return BitSet.valueOf(toLongArray(bitmap));
```

Comment:
Sorry, I had written up a long description of a thorny issue here which is the most important part of the review, which somehow got lost when "commenting" the review, I'm not sure I can write it out in full again, but tl;dr `BitSet` does weird things when constructed from `long[]` containing values >= `Integer.MAX_VALUE`, which `toLongArray()` can produce. The impact includes `BitSet#cardinality()` being correct, but inconsistent with iteration, idiomatic iteration loops throwing exceptions and so on. This can be demonstrated with this code:

```java
            long[] longArray = new long[(Integer.MAX_VALUE >>> 6) + 2];
            longArray[longArray.length - 1] = -1;
            longArray[0] = -1;
            BitSet bitSet = BitSet.valueOf(longArray);
            assertEquals(128, bitSet.cardinality());
            int computedCardinality = 0;
            int next = 0;
            while (true) {
                // will throw at the next iteration after producing -2147483648
                next = bitSet.nextSetBit(next + 1);
                if (next == -1) {
                    break;
                }
                computedCardinality++;
            }
            // will not get here, but this would fail no matter how the loop works around what's above
            assertEquals(bitSet.cardinality(), computedCardinality);
```

This is apparently already known about: [JDK-8311905](https://bugs.openjdk.org/browse/JDK-8311905) but it doesn't appear to be easy to fix, so we need to avoid creating `BitSet`s which behave strangely by checking the last value in the bitmap is < Integer.MAX_VALUE. This can be done by checking the 16th bit of the last key in the `RoaringArray` is absent.

Reply:
That's context I wasn't aware of, super useful, thanks. I'll add a check for that to make sure we don't get into trouble. I thought about the signed values, but figured that if you're using `BitSet` it's for interoperability/speed and you're already under `Integer.MAX_VALUE`.

I went with the allocation/copy  because I wanted `valueOf` to be fast as possible. I'm using this code already with `MethodHandles` to get at the `bitmap` field in my own code. I can add a variant that uses `forEach(IntConsumer)`, it looks like that should be the next fastest approach available.

We're using [Daniel's approach for iterating over BitSet](https://lemire.me/blog/2018/02/21/iterating-over-set-bits-quickly/) because it's about three times faster than the example for [nextSetBit](https://docs.oracle.com/javase/8/docs/api/java/util/BitSet.html#nextSetBit-int-) even with the array copy (for our use case, our largest bitmap is 7,439 words long):

```
    public final void bitIndexes(IntConsumer consumer) {
        long[] words = toLongArray();
        for (int k = 0; k < words.length; ++k) {
            long word = words[k];
            while (word != 0) {
                long t = word & Long.lowestOneBit(word);
                int r = Long.numberOfTrailingZeros(t);
                consumer.accept(k * Long.SIZE + r);
                word ^= t;
            }
        }
    }
```

Reply:
For reference, you could speed that up quite a lot by replacing the call to lowestOneBit (which isn’t an intrinsic and used quite a few instructions to isolate the bit last time I checked) with just numberOfTrailingZeros, as they are equivalent, then replacing the xor with word &= (word - 1) to unset the lowest bit, which is compiled to blsr on x86.

on the review, yes. I think adding the check and throwing when there is a value too large would suffice.